code/datums/components/echolocation.dm

```diff
@@ -0,0 +1,91 @@
+/datum/component/echolocation
+	var/echo_range = 4
+	var/cooldown = 0
+
+/datum/component/echolocation/Initialize()
+	. = ..()
+	RegisterSignal(parent, COMSIG_ECHOLOCATION_PING, .proc/echolocate)
+	var/mob/M = parent
+	M.audiolocation = TRUE
+	var/mob/living/carbon/human/H = parent
+	var/datum/species/gehennite/G = H.dna.species
+	G.awaken(H)
+
+/datum/component/echolocation/proc/echolocate()
+	var/mob/living/carbon/human/H = parent
+	var/mutable_appearance/image_output
+	var/list/seen = oview(echo_range, H)
+	var/list/filtered = list()
+	var/list/turfs = list()
+	if(!(cooldown < world.time - 30))
+		return
+	cooldown = world.time
+	for(var/I in seen)
+		var/atom/A = I
+		if(!(A.type in SSoutputs.echo_blacklist) && !A.invisibility)
+			if(istype(I, /obj))
+				if(istype(A.loc, /turf))
+					filtered += I
+			if(istype(I, /mob/living))
+				filtered += I
+			if(istype(I, /turf/closed/wall))
+				turfs += I
+	for(var/F in filtered)
+		var/atom/S = F
+		if(GLOB.echo_images[S.type])
+			image_output = mutable_appearance(GLOB.echo_images[S.type].icon, GLOB.echo_images[S.type].icon_state, GLOB.echo_images[S.type].layer, GLOB.echo_images[S.type].plane)
+			image_output.filters = GLOB.echo_images[S.type].filters
+			realign_icon(image_output, S)
+		else
+			image_output = generate_image(S)
+			realign_icon(image_output, S)
+			if(!(SSoutputs.uniques[S.type]))
+				GLOB.echo_images[S.type] = image_output
+		for(var/D in S.datum_outputs)
+			var/datum/outputs/O = D
+			if(O.echo_override)
+				image_output = mutable_appearance(O.echo_override.icon, O.echo_override.icon_state, O.echo_override.layer, O.echo_override.plane)
+				realign_icon(image_output, S)
+		H.display_output(, image_output,, S.loc)
+	for(var/T in turfs)
+		image_output = generate_wall_image(T)
+		image_output.loc = T
+		H.display_output(, image_output,, T)
+
+/datum/component/echolocation/proc/generate_image(atom/input)
+	var/icon/I
+	if(SSoutputs.uniques[input.type])
+		I = getFlatIcon(input)
```

Comment:
can't you just get the mutable appearance of the object rather than calling getflaticon?

Reply:
well not for humans and walls and shit since they're made up of a bunch of MAs

Reply:
that goes for the one below as well, you're turning the result into a mutable apperance anyway, why not just copy the existing atoms appearance and apply the filter to it and then just make sure it appears on the curse layer

Reply:
because i need to use icon mapcolor proc which only works on icons sadly

Reply:
fyi for anyone else who wondered this, there's no equivalent of MapColors for MA's and if you set the MA.color it bleeds into the outline

Reply:
probably because it applies after the filter